### ISSUE TITLE
do copy with CudaOpenGLImageInteropResource alwasy returns `ErrorInvalidValue`

### ISSUE THREAD

**Issue by jcyuan** [kunzmi/managedCuda#125]
Hi basically i have 2 questions:

1, i use CUDA for ffmpeg to decode video frames, and from ffmpeg's doc, the `AVFrame->data` contains `CUdeviceptr` for NV12 (default format for CUDA decoding)'s 2 planes data,
then i tried to use `CudaOpenGLImageInteropResource ` and copy these 2 ptr into OpenGL texture,
but failed, it always returns me `ErrorInvalidValue`, not sure why,
here is the key snippet of code, not sure what the problem is:

(note: i have allocated the gl texture size as w, h*1.5 (because i have combined NV12's 2planes onto the same texture)
```c#
public override void CopyD2D(IntPtr data, ulong size)
    {
        lock (_lock)
        {
            if (_cudaRes == null || _gl == null)
                throw new NullReferenceException("GL or res not initialized");

            using (_context.EnsureCurrent())
            {
                _cuContext.PushContext();
                
                if (!_cudaRes.IsMapped)
                    _cudaRes.Map();
                
                var nv = _cudaRes.GetMappedArray2D(0, 0);
                
                var ptrArray = *(byte_ptrArray8*)data;
                var byteSize = Marshal.SizeOf<byte>();
                var nSize = nv.Width * (long)((long)nv.Height / 1.5) * byteSize;
                var err = DriverAPINativeMethods.SynchronousMemcpy_v2.cuMemcpyDtoA_v2(nv.CUArray, 0,
                    new CUdeviceptr((long)ptrArray[0]), nSize);
                CheckCudaErrors(err);  // alwasy returns ErrorInvalidValue.....
                
                var vSize = nv.Width * (long)((long)nv.Height / 1.5 * 0.5) * byteSize;
                err = DriverAPINativeMethods.SynchronousMemcpy_v2.cuMemcpyDtoA_v2(nv.CUArray, nSize,
                    new CUdeviceptr((long)ptrArray[1]), vSize);
                CheckCudaErrors(err);
                
                _cudaRes.UnMap();

                _cuContext.PopContext();
            }

            _texDirty = true;
        }
    }
```
what am i doing wrong?  :(

2, about the license, i found that it's GPL license, if i need to use this lib in my close-source software, what should i do? if with commercial license, what is the price please?

thanks~

**Comment by kunzmi** [comment]
Hi,

for 1) where exactly is the exception thrown? How do you initialize the OpenGL texture? Are you sure that it can be mapped to CUArray? Another thing that I don't think that it can be correct is your copy method: you are using a 1D copy function for a 2D CUArray, which should also result in the mentioned ```ErrorInvalidValue```. Why not use the copy methods coming with the CudaArray2D class, something like ```nv.CopyFromHostToThis(IntPtr aHostSrc, SizeT aElementSizeInBytes)``` or ```nv.CopyData(CUDAMemCpy2D aCopyParameters)```?

for 2) please contact me by mail: [managedcuda@articimaging.eu](mailto:managedcuda@articimaging.eu)

Michael

**Comment by jcyuan** [comment]
> If it is the copy that fails, then it is for sure that it is because you use the 1D copy function for a 2D CUArray. You can either use directly `DriverAPINativeMethods.SynchronousMemcpy_v2.cuMemcpy2D_v2` or its wrapped call in `nv.CopyData(CUDAMemCpy2D aCopyParameters)` and provide the custom offsets in the copy parameters struct. Check [CudaArray2D.cs](https://github.com/kunzmi/managedCuda/blob/8ed144b3cc172a042e37b8f149ca9037a0801827/ManagedCUDA/CudaArray2D.cs) on how to fill the copy parameters struct.

thanks so much, let me have a try, will reply later about my result~

**Comment by jcyuan** [comment]
even a simple memory block will faill too, this is really frastrated.......  :(


```c#
_bufferPtr = _cudaContext.Value.AllocateMemory(length);
_destDeviceVar = new CudaDeviceVariable<byte>(_bufferPtr);

public override int CopyFrom(MediaFrame source)
    {
        _cudaContext.Value.PushContext();
        
        var offset = 0ul;
        for (uint i = 0; i < 8; i++)
        {
            if (vSrc.Pointer->data[i] == null)    //  data is nv planes (2 CUdevicePtr)
                continue;
            
            var plane = ffmpeg.av_frame_get_plane_buffer(vSrc.Pointer, (int)i);  //  get plane info
            var srcDevicePtr = new CUdeviceptr((long)plane->data);   // data is a CUdevicePtr
            
            _destDeviceVar.CopyToDevice(srcDevicePtr, 0, offset, plane->size);   // use device var copy to device with offset, but fail... the same result `ErrorInvalidValue`, i have to idea what to do...

            offset += plane->size;
        }
        
        _cudaContext.Value.PopContext();

        return (int)_bufferLength;
    }

![image](https://github.com/kunzmi/managedCuda/assets/850854/cd6cdf3d-f926-4843-ad2b-04634660172f)

could you help please... this is even not a interop, but a really simple GPU allocation & copy ffmpeg CUdevicePtr....

**Comment by kunzmi** [comment]
Having only small pieces of code, I can only guess what might be wrong...
MemCpy returns ErrorInvalidValue if you provide input parameters that the CUDA api cannot handle, basically this means that one of the device pointers is not a correct device pointer or that the size doesn't match. So in order to understand why Cuda is unhappy about your parameters, you'd need to debug a bit further to see if the values are what you think they should be:
```
//Create a CudaDeviceVariable<byte> from the CUdevicePtr coming from FFMPEG:
//This gathers the size of the allocation from the CUDA API, if this fails, plane->data is not a CUdevicePtr!
var testDevVar = new CudaDeviceVariable<byte>((long)plane->data); 
//check size, is it correct?
var testSize = testDevVar.SizeInBytes;
//copy data to host, check if data is what is expected:
byte[] testData = testDevVar;
```

and before the CopyToDevice-call, check that ```offset + plane->size``` is smaller than ```length```

**Comment by jcyuan** [comment]
> var testDevVar = new CudaDeviceVariable<byte>((long)plane->data); 
> //check size, is it correct?
> var testSize = testDevVar.SizeInBytes;
> //copy data to host, check if data is what is expected:
> byte[] testData = testDevVar;

thanks man!

yes that pointer is not valid, i'm not sure why...

step into this method,
![image](https://github.com/kunzmi/managedCuda/assets/850854/738d11bb-544f-4cd5-8a60-e91b931fa4cf)

and from here, it pops that the `cuMemGetAddressRange_v2` calling was failed.
![image](https://github.com/kunzmi/managedCuda/assets/850854/a6f0725e-5847-479f-9036-1b23485df29e)

this is werid, from ffmpeg's doc, it says:
![image](https://github.com/kunzmi/managedCuda/assets/850854/24472a4c-d3f6-40f1-a559-8779c8cabe71)

**Comment by jcyuan** [comment]
@kunzmi sir, i have solved this problem, it's quite simple, because i didn't push the context from ffmpeg  which contains the memory to current.

and i found a problem, the `CudaContext` class does not have a constructor which accepts an external context pointer.

**Comment by kunzmi** [comment]
yes, if ffmpeg library uses a different CudaContext, the allocations are not valid in your application...
ManagedCuda doesn't have a constructor with a pointer to external context as this is not a normal use case. What you can do instead is, making the context of ffmpeg current and fetch that context with the constructor ```CudaContext(int deviceId, bool createNew)``` where you set ```createNew = false```.
You likely don't have control over the ffmpeg context creation settings, for example binding it to OpenGL, etc...
Because sharing contexts along libraries is messy, Nvidia created the concept of a ```PrimaryContext```, so instead of using a standard ```CudaContext```, go with a PrimaryContext instead, it is always the same in one process and it is always bound to OpenGL, so likely ```PrimaryContext``` is what you are looking for.

**Comment by jcyuan** [comment]
Thanks so much.

I'm a bit unclear about the relationship between CUDA and OpenGL contexts. Is it the case that only specially created contexts can interop with OpenGL contexts? Or is it that after creating a CUDA context, there are additional API calls required to bind it with an OpenGL context?

Another question concerns the creation of a primary context and its association with FFmpeg's context. Does this mean that first I have to call the native API to set the FFmpeg's context as current, and then creating a primary context will automatically wrap it with the current context? and then, i can use this primary context everywhere?

sorry for bothering again.

**Comment by jcyuan** [comment]
ah! i assume you mean if i create the primary context before lunching the ffmpeg logic, later it will use this primary context to create sutffs?

**Comment by jcyuan** [comment]
Hi, @kunzmi  sorry for bothering again, 
actually now my CUDA -> Opengl Copy is working, but i got black screen, (shader works because i can't use some debug output, just the texture is black)

from what you said, i guess i need to use a special `cuContext` (or `PrimaryContext`), but if so, how can i make ffmpeg's `cuContext` current? i'm using `createNew = false` to get the ffmpeg's context for doing those `map/copy` actions, it works fine with `Success` returned, but just got black texture...

basically, 
1, make my OpenGL context as current (for binding opengl resources)
2, make ffmpeg's cuContext current (for map resource and copy to opengl)
3, so in this case, where to use that `PrimaryContext` which has OpenGL interopbility?

thanks again.

**Comment by jcyuan** [comment]
solved~ thanks so much.  :)  ❤
